neo/io/nwbio.py

```diff
@@ -0,0 +1,823 @@
+                        if elec_meta["device"]["name"] in devices:
+                            device = devices[elec_meta["device"]["name"]]
+                        else:
+                            device = nwbfile.create_device(**elec_meta["device"])
+                            devices[elec_meta["device"]["name"]] = device
+                        elec_meta.pop("device")
+                        electrodes[elec_meta["name"]] = nwbfile.create_icephys_electrode(
+                            device=device, **elec_meta
+                        )
+        return electrodes
+
+    def _write_segment(self, nwbfile, segment, electrodes):
+        # maybe use NWB trials to store Segment metadata?
+        for i, signal in enumerate(chain(segment.analogsignals, segment.irregularlysampledsignals)):
+            assert signal.segment is segment
+            if not signal.name:
+                signal.name = "%s : analogsignal%d" % (segment.name, i)
+            self._write_signal(nwbfile, signal, electrodes)
+
+        for i, train in enumerate(segment.spiketrains):
+            assert train.segment is segment
+            if not train.name:
+                train.name = "%s : spiketrain%d" % (segment.name, i)
+            self._write_spiketrain(nwbfile, train)
+
+        for i, event in enumerate(segment.events):
+            assert event.segment is segment
+            if not event.name:
+                event.name = "%s : event%d" % (segment.name, i)
+            self._write_event(nwbfile, event)
+
+        for i, epoch in enumerate(segment.epochs):
+            if not epoch.name:
+                epoch.name = "%s : epoch%d" % (segment.name, i)
+            self._write_epoch(nwbfile, epoch)
+
+    def _write_signal(self, nwbfile, signal, electrodes):
+        hierarchy = {'block': signal.segment.block.name, 'segment': signal.segment.name}
+        if "nwb_neurodata_type" in signal.annotations:
+            timeseries_class = get_class(*signal.annotations["nwb_neurodata_type"])
+        else:
+            timeseries_class = TimeSeries  # default
+        additional_metadata = {name[4:]: value
+                               for name, value in signal.annotations.items()
+                               if name.startswith("nwb:")}
+        if "nwb_electrode" in signal.annotations:
+            electrode_name = signal.annotations["nwb_electrode"]["name"]
+            additional_metadata["electrode"] = electrodes[electrode_name]
+        if timeseries_class != TimeSeries:
+            conversion, units = get_units_conversion(signal, timeseries_class)
+            additional_metadata["conversion"] = conversion
+        else:
+            units = signal.units
+        if isinstance(signal, AnalogSignal):
+            sampling_rate = signal.sampling_rate.rescale("Hz")
+            tS = timeseries_class(
+                name=signal.name,
+                starting_time=time_in_seconds(signal.t_start),
+                data=signal,
```

Comment:
With implementation we handle the signal in memmory.
We also should have later a mechanism that write a AnalogSignal in a chunk mode to avoid a full load in memory otherwise it will be impossible to write big files to nwb.

Maybe we can open issue for this to not forget it.

Reply:
this should go into the collective improvements issue for nwb

Reply:
added to #1013